Added API doc for myaccount phone /challenge and /verify endpoints (#3105)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/myaccount/index.md

}
```

> ### Send Phone Challenge

Reviewer: we should add these titles on top, under MyAccount Operations

| ---------- | ------ | ------------------------------------------------- |
| `id` | String | The id of the phone factor. Found in the response when a new phone number is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`)  |

#### Request query parameters

N/A

#### Request body

This request requires the `method` property as its request body. An optional boolean `retry` property can be used to indicate resend challenge when its value is set to `true`.

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `method`  | String | The method with which the challenge should be sent, valid values are `SMS` and `CALL` |
| `retry`  | Boolean | An optional property that indicates whether this is a normal challenge or retry |

#### Response body

Returns an empty response with an HTTP 200 status code.

#### Usage example

##### Request

This request sends a verification code by SMS to the phone number represented by `id`. The request is a normal phone challenge, not a retry.

```bash

> curl -XPOST 'https://${yourOktaDomain}/myaccount/phones/{id}/challenge' -H 'Authorization: SSWS {token}' -H 'Content-Type: application/json' --data '{

Reviewer: token is not SSWS, is Bearer

Author: Good catch, I copied from above and forgot to update it. Thanks.

}
```

### Send Phone Challenge

> <ApiOperation method="post" url="/idp/myaccount/phones/{id}/challenge"/>

Reviewer: show we doc scopes here too? my.account.phone.manage/read

> Send a phone challenge via one of the two methods: `SMS` or `CALL`. This API handles resend challenge (retry) as well.

Reviewer 2: ```suggestion
Send a phone challenge using one of two methods: `SMS` or `CALL`. The MyAccount API can also handle a resend challenge (retry).
```

> Upon a successful challenge, the user will receive a verification code via either `SMS` or `CALL`. The verification code can be used to verify the phone number by calling `POST /idp/myaccount/phones/{id}/verify` endpoint. The verification code would expire in 5 minutes.

Reviewer 2: ```suggestion
Upon a successful challenge, the user receives a verification code by `SMS` or `CALL`. Send a `POST` request to the `/idp/myaccount/phones/{id}/verify` endpoint to use the verification code to verify the phone number. The verification code expires in 5 minutes.
```

> This endpoint is supposed to be called no more than once every 30 seconds.

Author: The request will not be successful and the user will get a 429 (Too Many Requests) response. Let me know if I should include this in the doc, or drop the "note" format, thanks.

> This endpoint is supposed to be called no more than once every 30 seconds.

Reviewer 2: @hanyuanzhang-okta What about this?

> **Note:** Sending requests to the `/idp/myaccount/phones/{id}/verify` endpoint more often than once very 30 seconds returns a 429 (Too Many Requests) error.

#### Request path parameters

> `id`: the id of the phone factor, which can be found in the response when a new phone is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`).

Reviewer 2: We should put these parameters in table format. Is `id` a string type? Maybe instead of including the calls in the description, we could link to them in the docs?
```suggestion
| Parameter  | Type   | Description                                       |
| ---------- | ------ | ------------------------------------------------- |
| `id` | String | The id of the phone factor. Found in the response when a new phone is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`)  |
```

Author: Good point, will change it to a table. `id` is of string type. All paths variables and request properties in /challenge and /verify are either string or boolean.

#### Request query parameters

N/A

#### Request body

> This API requires the `method` property as its request body. An optional boolean `retry` property can be used to indicate resend challenge when its value is set to `true`.

Reviewer 2: ```suggestion
This request requires the `method` property as its request body. An optional boolean `retry` property can be used to indicate resend challenge when its value is set to `true`.
```

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `method`  | String | The method with which the challenge should be sent, valid values are `SMS` and `CALL` |
| `retry`  | Boolean | An optional property that indicates whether this is a normal challenge or retry |

#### Response body

> Returns an empty response with http status code of 200.

Reviewer 2: ```suggestion
Returns an empty response with an HTTP 200 status code.
```

#### Usage example

##### Request

> This request would result in a verification code to be sent to the phone number represended by the `id` via SMS. This is a normal phone challenge, not a retry.

Reviewer 2: ```suggestion
This request sends a verification code by SMS to the phone number represented by `id`. This is a normal phone challenge, not a retry.
```

```bash
curl -XPOST 'https://${yourOktaDomain}/myaccount/phones/{id}/challenge' -H 'Authorization: SSWS {token}' -H 'Content-Type: application/json' --data '{
     "method": "SMS",
     "retry": false
 }'
```

### Verify Phone Challenge

<ApiOperation method="post" url="/idp/myaccount/phones/{id}/verify"/>

> Verify the phone number with the verification code that the user received via `SMS` or `CALL`. The phone number would be active upon a successful verification.

Reviewer 2: ```suggestion
Verify the phone number with the verification code that the user receives via `SMS` or `CALL`. The phone number is active upon a successful verification.
```

#### Request path parameters

`id`: the id of the phone factor, which can be found in the response when a new phone number is added successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`).

#### Request query parameters

N/A

#### Request body

This API requires the `verificationCode` property as its request body.

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `verificationCode`  | String | A 6-digit verification code that the user receives via `SMS` or `CALL` |

#### Response body

Returns an empty response with http status code of 204.

#### Usage example

##### Request

> This request would verify the phone number represented by the `id` with the verification code of 796672.

Reviewer 2: ```suggestion
The following request verifies the phone number represented by `id` with a `verificationCode` of 796672.
```

}
```

### Send Phone Challenge

<ApiOperation method="post" url="/idp/myaccount/phones/{id}/challenge"/>

Sends a phone challenge using one of two methods: `SMS` or `CALL`. This request can also handle a resend challenge (retry).

Upon a successful challenge, the user receives a verification code by `SMS` or `CALL`. Send a `POST` request to the `/idp/myaccount/phones/{id}/verify` endpoint to use the verification code to verify the phone number. The verification code expires in 5 minutes.

> > **Note:** Do not call the `/idp/myaccount/phones/{id}/challenge` endpoint more often than once every 30 seconds.

Reviewer 2: ```suggestion
> **Note:** Don't call the `/idp/myaccount/phones/{id}/challenge` endpoint more often than once every 30 seconds.
```

#### Required scope and role

#### Request path parameters

| Parameter  | Type   | Description                                       |
| ---------- | ------ | ------------------------------------------------- |
| `id` | String | The id of the phone factor. Found in the response when a new phone number is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`)  |

#### Request query parameters

N/A

#### Request body

This request requires the `verificationCode` property as its request body.

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `verificationCode`  | String | A 6-digit verification code that the user receives via `SMS` or `CALL` |

#### Response body

Returns an empty response with an HTTP 204 status code.

Disabling the factor type of the corresponding method on the org returns a 403 FORBIDDEN with error code E0000038.
Disabling the IDP_MY_ACCOUNT_API feature flag returns a 401 UNAUTHORIZED with error code E0000015.
Passing an invalid `id` returns a 404 NOT FOUND with error code E0000008.

> Passing an invalid `verificationCode` returns a 400 BAD REQUEST with error code E0000001.

Author: This happens when we could not find a matching cached challenge from redis based on `factorId` and `verificationCode`.

> Passing a valid `verificationCode` again (verifying a phone number that has been verified) results a no-op. In an extremely rare case, verifying a phone number that has been verified returns a 409 BAD CONFLICT with error code E0000157.

Author: The second half is an edge case that could happen in a short window of a few milliseconds, when the transaction expired, but the cached challenge is not (see https://github.com/okta/okta-core/blob/master/components/web/end-user-app/web/src/main/java/com/okta/api/internal/mediation/myaccount/IdpMyAccountPhoneMediationServiceImpl.java#L314). Should ask user to call /challenge and /verify again. Let me know how do you want this to be handled.

| `id` | String | The id of the phone factor. Found in the response when a new phone number is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`)  |

#### Request query parameters

N/A

#### Request body

This request requires the `method` property as its request body. An optional boolean `retry` property can be used to indicate resend challenge when its value is set to `true`.

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `method`  | String | The method with which the challenge should be sent, valid values are `SMS` and `CALL` |
| `retry`  | Boolean | An optional property that indicates whether this is a normal challenge or retry |

#### Response body

Returns an empty response with an HTTP 200 status code.

Passing an invalid `method` returns a 400 BAD REQUEST with error code E0000001.

Disabling the factor type of the corresponding method on the org returns a 403 FORBIDDEN with error code E0000038.

Passing an invalid `id` returns a 404 NOT FOUND with error code E0000008.

> Call providers failing to send the out of band OTP challenge returns 500 with error code E0000138.

Author: This is not a client's side error, let me know if it should be mentioned here.

| `id` | String | The id of the phone factor. Found in the response when a new phone number is created successfully (`POST /idp/myaccount/phones`) or phone(s) is retrieved (`GET /idp/myaccount/phones`)  |

#### Request query parameters

N/A

#### Request body

This request requires the `verificationCode` property as its request body.

| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `verificationCode`  | String | A 6-digit verification code that the user receives via `SMS` or `CALL` |

#### Response body

Returns an empty response with an HTTP 204 status code.

Disabling the factor type of the corresponding method on the org returns a 403 FORBIDDEN with error code E0000038.

Disabling the `IDP_MY_ACCOUNT_API` feature flag on the org returns a 401 UNAUTHORIZED with error code E0000015.

Passing an invalid `id` returns a 404 NOT FOUND with error code E0000008.

Passing an invalid `verificationCode` returns a 400 BAD REQUEST with error code E0000001.

Passing a valid `verificationCode` again (verifying a phone number that has been verified) results a no-op.

> In an extremely rare case, verifying a phone number when the challenge is about to expire returns a 409 CONFLICT with error code E0000157.

Author: This is an edge case that could happen in a short window of a few to dozens of milliseconds, when the transaction expired, but the cached challenge is not (see https://github.com/okta/okta-core/blob/master/components/web/end-user-app/web/src/main/java/com/okta/api/internal/mediation/myaccount/IdpMyAccountPhoneMediationServiceImpl.java#L314). Should ask user to call /challenge and /verify again. Let me know how do you want this to be handled.
